Wip/hughsie/generic firmware

hughsie opened the pull request:
For comments, little tested.

superm1 commented on src/fu-tool.c:
```diff
@@ -1314,6 +1314,51 @@ fu_util_monitor (FuUtilPrivate *priv, gchar **values, GError **error)
+		}
+		str = fu_firmware_to_string (firmware);
+		g_print ("%s", str);
+	}
```
So the basic thought being you pass this firmware through all known types and then see how it gets interpreted if parsed by the different parsers?

I think I'd rather it be an extra argument of the type of parser to run it through.  Very few firmware will it make sense to just be parsed by a bunch of different parsers.

superm1 replied:
Furthermore if you run with this idea to only run it through one parser, I think it would be good to add another argument to `fwupdtool` for `get-firmware-parsers` that returns a list of known `GTypes` that the bash-completion could potentially use (similar to how `get-plugins` works).

And if you don't pass it into this function as an argument then there can be a prompt for numeric that lets you pick which parser to use.

superm1 commented on src/fu-tool.c:
```diff
@@ -1602,6 +1647,12 @@ main (int argc, char *argv[])
 		     _("Update the stored metadata with current contents"),
 		     fu_util_verify_update);
+	fu_util_cmd_array_add (cmd_array,
+		     "fwdump",
```
how about `firmware-parse` instead?

superm1 commented on src/fu-tool.c:
```diff
@@ -1602,6 +1647,12 @@ main (int argc, char *argv[])
+		     "FILENAME",
+		     /* TRANSLATORS: command description */
+		     _("Parse and show details about a firmware file"),
+		     fu_util_fwdump);
```
Need to add a bash completion snippet to the fwupdtool file.

superm1 commented:
I added some fixups for my comments, hope that's okay.  It doesn't seem to work before or after though (crashes initializing the gtype still).

superm1 commented:
OK this seems to work for me now.

hughsie commented:
Great, thanks for helping with this. I've fixed up a few more things and squashed it down. I did also ponder a `fwupdtool convert dfu raw foo.dfu bar.raw` on the logic the one in dfu-tool is somewhat limited now.

superm1 commented:
> Great, thanks for helping with this. I've fixed up a few more things and squashed it down. I did also ponder a fwupdtool convert dfu raw foo.dfu bar.raw on the logic the one in dfu-tool is somewhat limited now.

I think a convert function makes fine sense as a follow up sure.
